R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,229 @@
+    set(LIBR_LIB_DIR "${LIBR_HOME}/lib" CACHE PATH "R lib directory")
+  endif()
+
+# detection for UNIX & Win32
+else()
+
+  # attempt to find R executable
+  if(NOT LIBR_EXECUTABLE)
+
+    # CRAN may run RD CMD CHECK instead of R CMD CHECK,
+    # which can lead to this infamous error:
+    # 'R' should not be used without a path -- see par. 1.6 of the manual
+    find_program(
+      LIBR_EXECUTABLE
+      NO_DEFAULT_PATH
+      HINTS "${CMAKE_CURRENT_BINARY_DIR}" "/usr/bin" "/usr/lib/"
+      NAMES R R.exe
+    )
+
+    if(LIBR_EXECUTABLE MATCHES ".*lightgbm\\.Rcheck.*")
+      message(FATAL_ERROR "If you are seeing this error, it means you are running R CMD check and R is using '${LIBR_EXECUTABLE}'. Edit src/cmake/modulesFindLibR.cmake and add your R path to HINTS near this error")
+    endif()
+  endif()
+
+  if(UNIX)
+
+    if(NOT LIBR_EXECUTABLE)
+      message(FATAL_ERROR "Unable to locate R executable.\
+        \nEither add its location to PATH or provide it through the LIBR_EXECUTABLE cmake variable")
+    endif()
+
+    # ask R for the home path
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--vanilla" "-e" "cat(R.home())"
+      OUTPUT_VARIABLE LIBR_HOME
+    )
+
+    # ask R for the include dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('include'))"
+      OUTPUT_VARIABLE LIBR_INCLUDE_DIRS
+    )
+
+    # ask R for the lib dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('lib'))"
+      OUTPUT_VARIABLE LIBR_LIB_DIR
+    )
+
+  # Windows
+  else()
+
+    # ask R for R_HOME
+    if(LIBR_EXECUTABLE)
+      execute_process(
+        COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(normalizePath(R.home(),winslash='/'))"
+        OUTPUT_VARIABLE LIBR_HOME
+      )
+    endif()
+
+    # if R executable not available, query R_HOME path from registry
+    if(NOT LIBR_HOME)
+
+      # Try to find R's location in the registry
+      # ref: https://cran.r-project.org/bin/windows/base/rw-FAQ.html#Does-R-use-the-Registry_003f
+      get_filename_component(
+        LIBR_HOME
+        "[HKEY_LOCAL_MACHINE\\SOFTWARE\\R-core\\R\\${CMAKE_R_VERSION};InstallPath]"
+        ABSOLUTE
+      )
+
+      if(NOT LIBR_HOME)
+        get_filename_component(
+          LIBR_HOME
+          "[HKEY_CURRENT_USER\\SOFTWARE\\R-core\\R\\${CMAKE_R_VERSION};InstallPath]"
+          ABSOLUTE
+        )
+      endif()
+
+      if(NOT LIBR_HOME)
+        message(FATAL_ERROR "\nUnable to locate R executable.\
+          \nEither add its location to PATH or provide it through the LIBR_EXECUTABLE cmake variable")
+      endif()
+
+    endif()
+
+    # set exe location based on R_ARCH
+    if(NOT LIBR_EXECUTABLE)
+      set(LIBR_EXECUTABLE "${LIBR_HOME}/bin/${R_ARCH}/R.exe")
+    endif()
+
+    # set other R paths based on home path
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/bin/${R_ARCH}")
```

Comment:
Why can't we do the same for Windows?
https://github.com/jameslamb/LightGBM/blob/2858c1c4a0e1c91f564c0222b815e42f2dd1c0ec/R-package/src/cmake/modules/FindLibR.cmake#L125-L135

Reply:
do you mean "for non-Windows"? This block is in a part where  we are working with Windows.

I have seen that on Window's, R's source, shared library, and executables are always bundled together like this. On different Linux systems, R's installation follows the conventions of that OS and doesn't necessarily keep everything together.

For example on Debian (the [OS used by the Dockerfile we provide](https://github.com/jameslamb/LightGBM/blob/2858c1c4a0e1c91f564c0222b815e42f2dd1c0ec/docker/dockerfile-r#L2) I've seen this

```
-- LIBR_HOME [/usr/local/lib/R]
-- LIBR_EXECUTABLE [/usr/local/bin/R]
-- LIBR_INCLUDE_DIRS [/usr/local/lib/R/include]
-- LIBR_LIB_DIR [/usr/local/lib/R/lib]
```

Reply:
I mean, use code from the link https://github.com/jameslamb/LightGBM/blob/2858c1c4a0e1c91f564c0222b815e42f2dd1c0ec/R-package/src/cmake/modules/FindLibR.cmake#L125-L135
```
    # ask R for the include dir
    execute_process(
      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('include'))"
      OUTPUT_VARIABLE LIBR_INCLUDE_DIRS
    )

    # ask R for the lib dir
    execute_process(
      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('lib'))"
      OUTPUT_VARIABLE LIBR_LIB_DIR
    )
```
as a try to find dirs on Windows and macOS as well. At present this code is used only for Unix.

In other words, why don't we use `R.home('include')` and `R.home('lib')` on Windows and macOS?

Reply:
ohhhhhh ok. Yeah that's a good idea! Maybe I can move things around to reduce duplication if we do that. Push all the complexity back to R and out of my slapped-together `.cmake` script 😀

Reply:
@StrikerRUS ok I think I was able to take your suggestions and simplify this a lot! https://github.com/microsoft/LightGBM/pull/2901/commits/20f845255513a862b61ddd236b4bb6633a3ba697

Now each OS-specific block just tries to find `LIBR_EXECUTABLE`, and then re-uses that to get the other variables. This worked locally on my Mac (with both `Rscript build_r.R` and `R CMD CHECK`). Going to check on my Windows 10 laptop right now

Reply:
Ok I tested with Windows 10, compiling with Visual Studio 16 2019. I made a few small changes (https://github.com/microsoft/LightGBM/pull/2901/commits/4375584f43badfc8b83efe6f2c752700195c7c5e), but it works!